src/osd/ReplicatedBackend.cc

```diff
@@ -572,6 +655,7 @@ void ReplicatedBackend::op_commit(const ceph::ref_t<InProgressOp>& op)
     op->on_commit = 0;
     in_progress_ops.erase(op->tid);
   }
+  maybe_kick_pct_update();
```

Comment:
Might be missing something here, why can't we use a timestamp instead of a dedicated timer? 
We can introduce a last_pct_update timestamp which will be compared to, lazily, on the first attempted replica read.

Reply:
Because that check would be in the replica, not the primary.  We need the primary to propagate this bound so that the replicas actually have it.

Reply:
Seems like that would defeat the purpose -- you "need" an MOSDPGPCT when you receive an IO on an object with a log entry more recent than the current pct timestamp.  Waiting for a message round trip to the primary would be as bad as redirecting the client to the primary (which is what we do now).  I think the timer is a good middle ground between updating PCT on every IO and only piggy-backing it -- it allows us to ensure that an read on an object which hasn't seen an write in a configured number of seconds will succeed.

Reply:
> Waiting for a message round trip to the primary would be as bad as redirecting the client to the primary (which is what we do now) 

We shouldn't actually wait for the message to be replied, we can keep redirecting the client until the reply there. This behavior will allow us to serve replica reads at a _certain point_ unlike the exiting state where we must wait for the first write to the pg.
As I've seen it, paying the price of bouncing the "first" few read ops is considerable upon introducing a new timer.

Reply:
Approving as this comment shouldn't block we could start testing this PR.

Reply:
> > Waiting for a message round trip to the primary would be as bad as redirecting the client to the primary (which is what we do now)
> 
> We shouldn't actually wait for the message to be replied, we can keep redirecting the client until the reply there. This behavior will allow us to serve replica reads at a _certain point_ unlike the exiting state where we must wait for the first write to the pg. As I've seen it, paying the price of bouncing the "first" few read ops is considerable upon introducing a new timer.

I don't think I'm following.  Your suggestion is that we wait for the replica to bounce a replica-read to the primary to ask for an MOSDPGPCT?  It's not clear to me how that's an improvement on the timer.  It means that we need another message (MOSDPGPCTRequest) and would generally require at least one failed replica read before the replicas catch up.